docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,544 @@
+    - [Non-Goals / Future Work](#non-goals--future-work)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [Identified features from user stories](#identified-features-from-user-stories)
+    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      - [New API Types](#new-api-types)
+      - [Modifications required to existing API Types](#modifications-required-to-existing-api-types)
+      - [Behavioral Changes from v1alpha2](#behavioral-changes-from-v1alpha2)
+      - [Behaviors](#behaviors)
+        - [Create](#create)
+        - [Scale Up](#scale-up)
+        - [Scale Down](#scale-down)
+        - [Delete of the entire ControlPlane (kubectl delete controlplane my-controlplane)](#delete-of-the-entire-controlplane-kubectl-delete-controlplane-my-controlplane)
+        - [Cluster upgrade (using create-swap-and-delete)](#cluster-upgrade-using-create-swap-and-delete)
+        - [Control plane healthcheck](#control-plane-healthcheck)
+        - [Adoption of pre-v1alpha3 Control Plane Machines](#adoption-of-pre-v1alpha3-control-plane-machines)
+      - [Code organization](#code-organization)
+    - [Risks and Mitigations](#risks-and-mitigations)
+      - [etcd membership](#etcd-membership)
+  - [Design Details](#design-details)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+      - [Alpha -> Beta Graduation](#alpha---beta-graduation)
+    - [Upgrade Strategy](#upgrade-strategy)
+  - [Alternatives](#alternatives)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+The lexicon used in this document is described in more detail [here](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/reference/glossary.md).  Any discrepancies should be rectified in the main Cluster API glossary.
+
+### References
+
+[Kubernetes Control Plane Management Overview](https://docs.google.com/document/d/1nlWKEr9OP3IeZO5W2cMd3A55ZXLOXcnc6dFu0K9_bms/edit#)
+
+## Summary
+
+Management of the control plane is currently handled in Cluster API at the infrastructure provider level (although many providers have converged on using the same implementation), based on the special annotation of standard Cluster API machines which then changes the machine bootstrapping logic within a particular infrastructure provider.
+The purpose of this document is to instead treat machine-based control plane management as a first-class component that is part of the API and Cluster, where multiple Cluster API machines will belong to a single control plane object, with its own infrastructure independent provider. The control plane covered by this document is defined as consisting of the Kubernetes API server, scheduler, controller manager, DNS and proxy services, and the underlying etcd datastore.
+
+## Motivation
+
+### Goals
```

Comment:
I do not see in goals or non-goals any mention of upgrades of single node control planes where there is no external control plane endpoint (CPE). In a Slack conversation with @ncdc, I said my understanding was that single-node control planes without an external CPE would not be supported by any upgrade logic built into CAPI. I think it's worth calling that out explicitly.